File: Core/Python/dictionary.py
INITIAL_ITEMS_SIZE = 100

def hash(key, max):
  totalHash = 0
  prime = 37
  for char in key:
    totalHash += prime * (totalHash + ord(char))
  return totalHash % max

def getIndexOfKey(key, chainArr):
  for (i, _) in enumerate(chainArr):
    chainItem = chainArr[i]
    if (chainItem[0] == key):
      return i
  return -1

class Dictionary:

  __items = []

  __insertionCount = 0

  __collisionCount = 0

  def __init__(self, keyValuePairs):
    for _ in range(0, INITIAL_ITEMS_SIZE):
      self.__items.append([])
    if (keyValuePairs and len(keyValuePairs) > 0):
      self.__setKeyValuePairs(keyValuePairs)

  def __setKeyValuePairs(self, keyValuePairs):
    if (keyValuePairs and len(keyValuePairs) > 0):
      for pair in keyValuePairs:
        self.set(pair[0], pair[1])

  def set(self, key, value):
    hashIndex = hash(key, len(self.__items) - 1)
    self.__insertionCount += 1
    newPair = [key, value, self.__insertionCount]
    if (len(self.__items[hashIndex]) > 0):
      # Collision occured.
      self.__collisionCount += 1
    if (self.__collisionCount > 10):
      self.__collisionCount = 0
      for _ in range(0, len(self.__items)):
        self.__items.append([])
    existingIndex = getIndexOfKey(key, self.__items[hashIndex])
    if (existingIndex > -1):
      self.__items[hashIndex][existingIndex] = newPair
    else:
      self.__items[hashIndex].append(newPair)

  def get(self, key):
    hashIndex = hash(key, len(self.__items) - 1)
    chain = self.__items[hashIndex]
    for item in chain:
      if (item[0] == key):
        return item[1]
    return None

  def has(self, key):
    return self.get(key) is not None

  def entries(self):
    chainEntries = []
    entries = filter(lambda chain: len(chain) > 0, self.__items)
    for chain in entries:
      for item in chain:
        chainEntries.append(item)
    return sorted(chainEntries, key=lambda arr: arr[2])

  def __str__(self):
    retStr = '{ \n'
    entries = self.entries()
    for (index, entry) in enumerate(entries):
      if (type(entry[1]) is str):
        retStr += "  \"" + str(entry[0]) + "\" : \"" + str(entry[1]) + "\""
      else:
        retStr += "  \"" + str(entry[0]) + "\" : " + str(entry[1])
      if (index < (len(entries) - 1)):
        retStr += ",\n"
    retStr += '\n}'
    return retStr

newDict = Dictionary([["hello", 89], ["what?", "who!"], ["Stamm", None]])
newDict.set("happy", "Good feeling!")
newDict.set("happy", None)
newDict.set("happy", "Not none!!")
print(newDict.entries())

File: Core/Swift/dictionary.swift
extension Character {
  func getCharCode() -> Int {
    let charStr = String(self);
    let charScalars = charStr.unicodeScalars;
    return Int(charScalars[charScalars.startIndex].value)  
  }
}

let INITIAL_ITEMS_SIZE = 100;

// Our hashing function...
func hash(key: String, maxNumber: Int) -> Int {
  var hashTotal = 0;
  let primeMultiplier = 37;
  for char in key {
    hashTotal = hashTotal + (primeMultiplier * (char.getCharCode() + hashTotal));
  }
  return hashTotal % maxNumber;
}

func getIndexOfKey(key: String, chainArr: [[Any]]) -> Int {
  for (i, _) in chainArr.enumerated() {
    let chainItem = chainArr[i];
    if (chainItem[0] as! String == key) {
      return i;
    }
  }
  return -1;
}

class Dictionary {

  private var items: [[[Any]]] = Array(repeating: [], count: INITIAL_ITEMS_SIZE)

  private var collisionCount = 0;

  private var insertionCount = 0;

  init() { }

  init(fromKeyValuePairs keyValuePairs: [[Any]]) {
    if (keyValuePairs.count > 0) {
      self.setKeyValuePairs(keyValuePairs: keyValuePairs);
    }
  }

  private func setKeyValuePairs(keyValuePairs: [[Any]]) -> Void {
    for keyValuePair in keyValuePairs {
      self.set(key: keyValuePair[0] as! String, value: keyValuePair[1]);
    }
  }

  public func set(key: String, value: Any) -> Void {
    let hashIndex = hash(key: key, maxNumber: self.items.count - 1);
    self.insertionCount += 1;
    let newPair = [key, value, self.insertionCount];
    if (self.items[hashIndex].count > 1) {
      // We are colliding.
      self.collisionCount += 1;
    }
    if (self.collisionCount > 10) {
      // Double the array size, we are colliding a lot.
      self.collisionCount = 0;
      self.items.append(contentsOf: Array(repeating: [], count: self.items.count))
    }
    let existingIndex = getIndexOfKey(key: key, chainArr: self.items[hashIndex]);
    if (existingIndex > -1) {
      self.items[hashIndex][existingIndex] = newPair;
    } else {
      self.items[hashIndex].append(newPair);
    }
  }

  public func get(key: String) -> Any? {
    let hashIndex = hash(key: key, maxNumber: self.items.count - 1);
    let chain = self.items[hashIndex];
    for pair in chain {
      if let pairKey = pair[0] as? String {
        if (pairKey == key) {
          return pair[1];
        }
      }
    }
    return nil;
  }

  public func remove(key: String) -> Void {
    let hashIndex = hash(key: key, maxNumber: self.items.count - 1);
    let chain = self.items[hashIndex];
    let existingIndex = getIndexOfKey(key: key, chainArr: chain);
    if (existingIndex > -1) {
      // we have an entry to delete.
      self.items[hashIndex].remove(at: existingIndex)
    }
  }

  public func has(key: String) -> Bool {
    return self.get(key: key) != nil;
  }

  public func entries() -> [[Any]] {
    var chainEntries: [[Any]] = [];
    let entries = self.items.filter { $0.count > 0 }
    for pair in entries {
      for item in pair {
        chainEntries.append(item);
      }
    }
    chainEntries.sort {
      let itemOne = $0 as [Any]
      let itemTwo = $1 as [Any]
      let orderOne = itemOne[2] as! Int
      let orderTwo = itemTwo[2] as! Int
      return orderOne < orderTwo;
    }
    return chainEntries;
  }

  public func toString() -> String {
    let entries = self.entries();
    var retStr = "{ \n";
    for (index, entry) in entries.enumerated() {
      let valueType = "\(type(of: entry[1]))"
      if (valueType == "String") {
        retStr += "  \"\(entry[0])\" : \"\(entry[1])\""
      } else {
        retStr += "  \"\(entry[0])\" : \(entry[1])"
      }
      if (index < entries.count - 1) {
        retStr += ",\n"
      }
    }
    retStr += "\n}"
    return retStr;
  }
}

let dict = Dictionary(fromKeyValuePairs: [["hey", "person!"]])
dict.remove(key: "hey");
dict.set(key: "mcklunky", value: 100);
print(dict.toString())
